### ISSUE TITLE
Improve Quantity and Unit system

### ISSUE THREAD

**Issue by peterNordin** [Hopsan/hopsan#1801]
### Issues with the current implementation
From the comments in PR #1793, the following issues have been identified.

- If no Quantity have been registered, and a unit that is a base-unit for two different quantities is given. The automatic Quantity selection does not work. E.g. Frequency and Angular Velocity both have rad/s as the base unit
- There is no easy way to change base unit for a quantity, then all loaded models must be updated, and likely all model code must also be. The latter is not possible to solve in hopsan. 
- Unit scaling is only known by HopsanGUI, Hopsan core only works with "base-units" (by design). But if unit scaling was known by the core, then maybe changing base unit would not be so complicated
- It would be nice (not sure I agree) to be able to use non-base-units in the component code. Technically you can use whatever unit you like, but it is confusing if the majority of components use un-scaled SI (or derived) base units, and then someone comes along and use something different. Ex: millimeter for a position instead of meter. But if the core new about unit scales, then maybe invalid (unit-wise) connections could be detected and prevented. See for example the FMI 2.0 specification.
- It is unclear how to set a unit or quantity on a variable/parameter in a component. And it is difficult to ensure that a correct value is used since a string is the parameter type. A user can write anything in the string. Example:  Ohm instead of ohm (actually one should write the quantity name Resistance). If Ohm is written it will not be possible change the unit scaling.

Out of battery will update later

**Comment by MarkusBagge** [comment]
Very nice summary. I didn't notice until now you had put it in.

**Comment by peterNordin** [comment]
Yes I was going to tell you about it when I was done editing it, but I never finished it, I will continue to improve and clarify this issue. And add some check boxes with tasks, and document decisions as they are made.

**Comment by peterNordin** [comment]
@robbr48 @MarkusBagge I have updated the issue description as best I can based on earlier discussions. Feel free to suggest improvements or request clarification. I have added my current opinion (which may change) to each decision item. @robbr48 You can add yours when you have time. when/if we agree I can add tasks.  
This will obviously take a very long time to complete, but one step at a time will lead to some improvements. 
Also requesting feedback (as comments) (if you have have time for it) from @petkr14 @magse @marthoch  I believe that you may have valuable input on this sort of thing.

**Comment by MarkusBagge** [comment]
Again, excellent break down! 
Regarding "Component developement/Point 1/Answer 2";
>But maybe the default display value should be possible to choose, maybe in the component appearance XML.

If the quantity system gets an overhaul the defaults set using Menu Tools->Options->Unit scaling should suffice? 

>Should we strive to make it possible to change the "base-unit" for a quantity (not needed if non-base-units are allowed)?
peter: No, old model will break, a hmf can be updated on load, but not someone else's code.

A "hmf" is a model, no? So then the problem boils down to custom written component code that uses "Frequency" rather than "rad/s". Code that uses "rad/s" will continue to work even though base unit is changed for the quantity I suppose.
Am thinking, just _in case_ not making changes will forever be haunting us in future, potentially creating more issues than already present, then Hopsan-release-notes.txt at least says
>- There is no guarantee that the included component models are working under all conditions,
  you must decide for yourself if you trust the simulation results.

Tiny comfort for those (few?) poor souls affected though.

>Should we use some other type then "char array/string" to set the Quantity / unit in the Core, and in that case what type? An alternative would be to check each unit/quantity against a list of allowed ones.

Or create overloaded methods that takes for e.g. ints (or objects) as unit identifiers? Example, using a namespace Units, to give auto-completion in QTCreator. 
```
addInputVariable("wc", "Friction speed", Units::rad_per_second, 1.,&mpwc);
```
As already mentioned, using predefined constants does not remove the need for using strings so that non-predefined units can still be used. Using an object - implementing exponent info according to FMI standard - provides means for using non-predefined derived units as well I guess. Working a bit on that one...
